File: samples/scala-eventsourced-counter/src/main/scala/com/example/domain/Counter.scala
package com.example.domain

import com.akkaserverless.scalasdk.eventsourcedentity.EventSourcedEntity
import com.akkaserverless.scalasdk.eventsourcedentity.EventSourcedEntityContext
import com.akkaserverless.scalasdk.SideEffect
import com.example
import com.google.protobuf.empty.Empty

// This class was initially generated based on the .proto definition by Akka Serverless tooling.
//
// As long as this file exists it will not be overwritten: you can maintain it yourself,
// or delete it so it is regenerated as needed.

/** An event sourced entity. */
class Counter(context: EventSourcedEntityContext) extends AbstractCounter {
  override def emptyState: CounterState = CounterState()

  override def increase(
      currentState: CounterState,
      increaseValue: example.IncreaseValue): EventSourcedEntity.Effect[Empty] =
    if (increaseValue.value < 0) {
      effects.error("Value must be a zero or a positive number")
    } else {
      effects
        .emitEvent(ValueIncreased(increaseValue.value))
        .thenReply(_ => Empty())
    }

  import com.akkaserverless.javasdk.impl.DeferredCallImpl

  override def increaseWithSideEffect(
      currentState: CounterState,
      increaseValue: example.IncreaseValue): EventSourcedEntity.Effect[Empty] =
    if (increaseValue.value < 0) {
      effects.error("Value must be a zero or a positive number")
    } else {
      effects
        .emitEvent(ValueIncreased(increaseValue.value))
        .thenAddSideEffect(_ =>
          SideEffect(components.counter.increase(example.IncreaseValue(context.entityId, increaseValue.value * 2))))
        .thenReply(_ => Empty())
    }

  override def decrease(
      currentState: CounterState,
      decreaseValue: example.DecreaseValue): EventSourcedEntity.Effect[Empty] =
    if (decreaseValue.value < 0) {
      effects.error("Value must be a zero or a positive number")
    } else {
      effects
        .emitEvent(ValueDecreased(decreaseValue.value))
        .thenReply(_ => Empty())
    }

  override def reset(currentState: CounterState, resetValue: example.ResetValue): EventSourcedEntity.Effect[Empty] =
    effects
      .emitEvent(ValueReset())
      .thenReply(_ => Empty())

  override def getCurrentCounter(
      currentState: CounterState,
      getCounter: example.GetCounter): EventSourcedEntity.Effect[example.CurrentCounter] =
    effects.reply(example.CurrentCounter(currentState.value))

  override def valueIncreased(currentState: CounterState, valueIncreased: ValueIncreased): CounterState =
    currentState.copy(value = currentState.value + valueIncreased.value)

  override def valueDecreased(currentState: CounterState, valueDecreased: ValueDecreased): CounterState =
    currentState.copy(value = currentState.value - valueDecreased.value)

  override def valueReset(currentState: CounterState, valueReset: ValueReset): CounterState =
    emptyState
}

File: samples/java-valueentity-counter-pojo/src/main/java/com/example/domain/Counter.java
/* This code was initialised by Akka Serverless tooling.
 * As long as this file exists it will not be re-generated.
 * You are free to make changes to this file.
 */

package com.example.domain;

import com.akkaserverless.javasdk.valueentity.ValueEntityContext;
import com.example.CounterApi;
import com.google.protobuf.Empty;

/**
 * A Counter represented as a value entity.
 */
public class Counter extends AbstractCounter {

    @SuppressWarnings("unused")
    private final String entityId;

    public Counter(ValueEntityContext context) {
        this.entityId = context.entityId();
    }

    @Override
    public CounterState emptyState() {
      return new CounterState(0);
    }

    @Override
    public Effect<Empty> increase(
        CounterState currentState, CounterApi.IncreaseValue command) {
      return effects()
              .updateState(currentState.increase(command.getValue()))
              .thenReply(Empty.getDefaultInstance());
    }

    @Override
    public Effect<Empty> decrease(
            CounterState currentState,
            CounterApi.DecreaseValue command) {

      return effects()
          .updateState(currentState.decrease(command.getValue()))
          .thenReply(Empty.getDefaultInstance());
    }
    
    @Override
    public Effect<Empty> reset(
        CounterState currentState, CounterApi.ResetValue command) {
      return effects()
          .updateState(emptyState())
          .thenReply(Empty.getDefaultInstance());
    }
    
    @Override
    public Effect<CounterApi.CurrentCounter> getCurrentCounter(
            CounterState currentState,
            CounterApi.GetCounter command) {
        CounterApi.CurrentCounter current =
                CounterApi.CurrentCounter.newBuilder()
                    .setValue(currentState.value).build();
        return effects().reply(current);
    }
}
